### PR TITLE
debit card expense support

### PR DESCRIPTION
**Pull request by labhvam5**

### CODE REVIEW

File: apps/mappings/utils.py [hunk 1]
```diff
@@ -39,7 +39,7 @@ def create_or_update_general_mapping(self, general_mapping: Dict):
             assert_valid('bank_account_id' in general_mapping and general_mapping['bank_account_id'],
                          'bank account id field is blank')
 
-        if general_settings.corporate_credit_card_expenses_object and \
+        if general_settings.corporate_credit_card_expenses_object == 'CREDIT CARD PURCHASE' and \
```
Comment (Shwetabhk) [inline apps/mappings/utils.py:42]:
```suggestion
        if general_settings.corporate_credit_card_expenses_object and \
              general_settings.corporate_credit_card_expenses_object not in ('BILL', 'DEBIT CARD EXPENSE'):
```

File: apps/quickbooks_online/models.py [hunk 2]
```diff
@@ -499,7 +499,7 @@ def create_qbo_expense(expense_group: ExpenseGroup):
         qbo_expense_object, _ = QBOExpense.objects.update_or_create(
             expense_group=expense_group,
             defaults={
-                'expense_account_id': general_mappings.qbo_expense_account_id,
+                'expense_account_id': general_mappings.qbo_expense_account_id if expense_group.fund_source == 'PERSONAL' else general_mappings.default_debit_card_account_id,
```
Comment (Shwetabhk) [inline apps/quickbooks_online/models.py:502]: This line seems too long, break it into next line

File: apps/quickbooks_online/tasks.py [hunk 3]
```diff
@@ -386,7 +389,7 @@ def __validate_expense_group(expense_group: ExpenseGroup, general_settings: Work
                 'message': 'CCC account mapping / Default CCC account mapping not found'
             })
 
-    if general_settings.corporate_credit_card_expenses_object != 'BILL' and expense_group.fund_source == 'CCC':
+    if general_settings.corporate_credit_card_expenses_object != 'BILL' and general_settings.corporate_credit_card_expenses_object != 'DEBIT CARD EXPENSE' and expense_group.fund_source == 'CCC':
```
Comment (Shwetabhk) [inline apps/quickbooks_online/tasks.py:392]: T
```suggestion
    if general_settings.corporate_credit_card_expenses_object not in ('BILL', 'DEBIT CARD EXPENSE') and expense_group.fund_source == 'CCC':
```

File: apps/quickbooks_online/tasks.py [hunk 4]
```diff
@@ -638,11 +651,17 @@ def create_qbo_expense(expense_group, task_log_id):
...
         qbo_connection = QBOConnector(qbo_credentials, expense_group.workspace_id)
+        
+        if expense_group.fund_source == 'PERSONAL':
+            if general_settings.auto_map_employees and general_settings.auto_create_destination_entity:
+                create_or_update_employee_mapping(expense_group, qbo_connection, general_settings.auto_map_employees)
 
-        if general_settings.auto_map_employees and general_settings.auto_create_destination_entity:
-            create_or_update_employee_mapping(expense_group, qbo_connection, general_settings.auto_map_employees)
-
+        if expense_group.fund_source == 'CCC':
```
Comment (Shwetabhk) [inline apps/quickbooks_online/tasks.py:659]: This should be an else case for condition on line 655

Comment (Shwetabhk) [reply apps/quickbooks_online/tasks.py:659]: Use else

File: apps/quickbooks_online/tasks.py [hunk 5]
```diff
@@ -759,9 +778,10 @@ def create_credit_card_purchase(expense_group: ExpenseGroup, task_log_id):
                 create_or_update_employee_mapping(expense_group, qbo_connection, general_settings.auto_map_employees)
         else:
             merchant = expense_group.expenses.first().vendor
-            get_or_create_credit_card_vendor(expense_group.workspace_id, merchant)
+            get_or_create_credit_card_or_debit_card_vendor(expense_group.workspace_id, merchant, False)
 
         with transaction.atomic():
+            print('this is credit card :(')
```
Comment (Shwetabhk) [inline apps/quickbooks_online/tasks.py:784]: remove print statement

File: apps/quickbooks_online/tasks.py [hunk 6]
```diff
@@ -638,11 +651,17 @@ def create_qbo_expense(expense_group, task_log_id):
...
 
-        if general_settings.auto_map_employees and general_settings.auto_create_destination_entity:
-            create_or_update_employee_mapping(expense_group, qbo_connection, general_settings.auto_map_employees)
-
+        else:
+            merchant = expense_group.expenses.first().vendor
+            get_or_create_credit_card_or_debit_card_vendor(expense_group.workspace_id, merchant, True)
+       
         with transaction.atomic():
+            print("hello this is qbo expense")
```
Comment (Sravanksk) [inline apps/quickbooks_online/tasks.py:664]: remove print statement

File: apps/quickbooks_online/models.py [hunk 7]
```diff
@@ -57,7 +57,10 @@ def get_expense_purpose(workspace_id, lineitem, category, workspace_general_sett
 
 def construct_private_note(expense_group: ExpenseGroup):
     expense = expense_group.expenses.first()
-    expense_type = 'Reimbursable' if expense_group.fund_source == 'PERSONAL' else 'Credit card'
+    workspace_general_settings = WorkspaceGeneralSettings.objects.get(workspace_id=expense_group.workspace_id)
+    expense_type = 'Reimbursable'
+    if (expense_group.fund_source == 'CCC'):
+        expense_type = 'Credit Card' if workspace_general_settings.corporate_credit_card_expenses_object == 'CREDIT CARD PURCHASE' else 'Debit card'
```
Comment (Sravanksk) [inline apps/quickbooks_online/models.py:63]: This is not correct. This will break for Journal Entry and Bills config

Comment (Sravanksk) [reply apps/quickbooks_online/models.py:63]: if (expense_group.fund_source == 'CCC'):
        expense_type = 'Debit Card' if workspace_general_settings.corporate_credit_card_expenses_object == 'DEBIT CARD EXPENSE' else 'Credit card'`

File: apps/workspaces/utils.py [hunk 8]
```diff
@@ -134,7 +134,7 @@ def create_or_update_general_settings(general_settings_payload: Dict, workspace_
     )
 
     if general_settings.map_merchant_to_vendor and \
-            general_settings.corporate_credit_card_expenses_object == 'CREDIT CARD PURCHASE':
+            general_settings.corporate_credit_card_expenses_object in ('CREDIT CARD PURCHASE','DEBIT CARD EXPENSE'):
```
Comment (Sravanksk) [inline apps/workspaces/utils.py:137]: indentation 
            `general_settings.corporate_credit_card_expenses_object in ('CREDIT CARD PURCHASE', 'DEBIT CARD EXPENSE'):`

Comment (Sravanksk) [reply apps/workspaces/utils.py:137]: no space after 'CREDIT CARD PURCHASE',

**Pull request merged**